Framework/Source/Graphics/RenderGraph/ResourceCache.cpp

```diff
@@ -35,27 +35,133 @@ namespace Falcor
         return SharedPtr(new ResourceCache());
     }
 
+    void ResourceCache::reset()
+    {
+        mNameToIndex.clear();
+        mResourceData.clear();
+        mExternalResources.clear();
+    }
+
     const std::shared_ptr<Resource>& ResourceCache::getResource(const std::string& name) const
     {
         static const std::shared_ptr<Resource> pNull;
-        const auto& resIt = mResources.find(name);
-        if (resIt == mResources.end())
+        const auto& it = mNameToIndex.find(name);
+        
+        // Search external resources if not found in render graph resources
+        if (it == mNameToIndex.end())
         {
-            logWarning("Can't find a resource named `" + name + "` in the resource cache");
-            return pNull;
+            auto extIt = mExternalResources.find(name);
+            if (extIt == mExternalResources.end())
+            {
+                logWarning("Can't find a resource named `" + name + "` in the resource cache");
+                return pNull;
+            }
+
+            return extIt->second;
         }
 
-        return resIt->second;
+        return mResourceData[it->second].pResource;
     }
 
-    void ResourceCache::addResource(const std::string& name, const std::shared_ptr<Resource>& pResource)
+    void ResourceCache::registerExternalResource(const std::string& name, const std::shared_ptr<Resource>& pResource)
     {
-        mResources[name] = pResource;
+        mExternalResources[name] = pResource;
     }
 
-    void ResourceCache::removeResource(const std::string& name)
+    void ResourceCache::removeExternalResource(const std::string& name)
     {
-        mResources.erase(name);
+        mExternalResources.erase(name);
     }
 
+    void ResourceCache::registerField(const std::string& name, const RenderPassReflection::Field& field, const std::string& alias)
+    {
+        if (mNameToIndex.count(name) > 0)
+        {
+            logWarning("ResourceCache::registerField: Field named " + name + " already exists. Ignoring operation.");
+            return;
+        }
+
+        bool addAlias = alias.empty() == false;
+        if (addAlias && mNameToIndex.count(alias) == 0)
+        {
+            logWarning("ResourceCache::registerField: Field named " + alias + " not found. Cannot add " + name + "as an alias. Creating new entry.");
+            addAlias = false;
+        }
+
+        // Add a new field
+        if (addAlias == false)
+        {
+            assert(mNameToIndex.count(name) == 0);
+            mNameToIndex[name] = (uint32_t)mResourceData.size();
+            mResourceData.push_back({ field, true, nullptr });
+        }
+        else
+        {
+            // Add alias
+            uint32_t index = mNameToIndex[alias];
+            mNameToIndex[name] = index;
+
+            // Merge fields, and overwrite previously unknown/unspecified fields with specified ones
+            RenderPassReflection::Field& cachedField = mResourceData[index].field;
```

Comment:
We should have a `mergeField` function, which returns a bool indicating whether the merge succeed or not